Mathlib/LinearAlgebra/Span.lean

```diff
@@ -157,82 +157,87 @@ theorem span_closure {s : Set M} : span R (AddSubmonoid.closure s : Set M) = spa
-for binary predicates. -/
-theorem span_induction₂ {p : M → M → Prop} {a b : M} (ha : a ∈ Submodule.span R s)
-    (hb : b ∈ Submodule.span R s) (mem_mem : ∀ x ∈ s, ∀ y ∈ s, p x y)
-    (zero_left : ∀ y, p 0 y) (zero_right : ∀ x, p x 0)
-    (add_left : ∀ x₁ x₂ y, p x₁ y → p x₂ y → p (x₁ + x₂) y)
-    (add_right : ∀ x y₁ y₂, p x y₁ → p x y₂ → p x (y₁ + y₂))
-    (smul_left : ∀ (r : R) x y, p x y → p (r • x) y)
-    (smul_right : ∀ (r : R) x y, p x y → p x (r • y)) : p a b :=
-  Submodule.span_induction ha
-    (fun x hx => Submodule.span_induction hb (mem_mem x hx) (zero_right x) (add_right x) fun r =>
-      smul_right r x)
-    (zero_left b) (fun x₁ x₂ => add_left x₁ x₂ b) fun r x => smul_left r x b
-
-/-- A dependent version of `Submodule.span_induction`. -/
-@[elab_as_elim]
-theorem span_induction' {p : ∀ x, x ∈ span R s → Prop}
+theorem span_induction {p : (x : M) → x ∈ span R s → Prop}
     (mem : ∀ (x) (h : x ∈ s), p x (subset_span h))
     (zero : p 0 (Submodule.zero_mem _))
     (add : ∀ x hx y hy, p x hx → p y hy → p (x + y) (Submodule.add_mem _ ‹_› ‹_›))
     (smul : ∀ (a : R) (x hx), p x hx → p (a • x) (Submodule.smul_mem _ _ ‹_›)) {x}
     (hx : x ∈ span R s) : p x hx := by
-  refine Exists.elim ?_ fun (hx : x ∈ span R s) (hc : p x hx) => hc
-  refine
-    span_induction hx (fun m hm => ⟨subset_span hm, mem m hm⟩) ⟨zero_mem _, zero⟩
-      (fun x y hx hy =>
-        Exists.elim hx fun hx' hx =>
-          Exists.elim hy fun hy' hy => ⟨add_mem hx' hy', add _ _ _ _ hx hy⟩)
-      fun r x hx => Exists.elim hx fun hx' hx => ⟨smul_mem _ _ hx', smul r _ _ hx⟩
+  let p : Submodule R M :=
+    { carrier := { x | ∃ hx, p x hx }
+      add_mem' := (Exists.elim · fun _ ha ↦ (Exists.elim · fun _ hb ↦ ⟨_, add _ _ _ _ ha hb⟩))
+      zero_mem' := ⟨_, zero⟩
+      smul_mem' := fun r ↦ (Exists.elim · fun _ hb ↦ ⟨_, smul r _ _ hb⟩) }
+  exact span_le (p := p) |>.mpr (fun y hy ↦ ⟨subset_span hy, mem y hy⟩) hx |>.elim fun _ ↦ id
+
+@[deprecated span_induction (since := "2024-10-10")]
+alias span_induction' := span_induction
+
+/-- An induction principle for span membership. This is a version of `Submodule.span_induction`
+for binary predicates. -/
+theorem span_induction₂ {p : (x y : M) → x ∈ span R s → y ∈ span R s → Prop}
+    (mem_mem : ∀ (x) (hx : x ∈ s) (y) (hy : y ∈ s), p x y (subset_span hx) (subset_span hy))
+    (zero_left : ∀ y hy, p 0 y (zero_mem _) hy) (zero_right : ∀ x hx, p x 0 hx (zero_mem _))
+    (add_left : ∀ x hx y hy z hz, p x z hx hz → p y z hy hz → p (x + y) z (add_mem hx hy) hz)
+    (add_right : ∀ x hx y hy z hz, p x y hx hy → p x z hx hz → p x (y + z) hx (add_mem hy hz))
+    (smul_left : ∀ (r : R) x hx y hy, p x y hx hy → p (r • x) y (smul_mem _ r hx) hy)
+    (smul_right : ∀ (r : R) x hx y hy, p x y hx hy → p x (r • y) hx (smul_mem _ r hy))
+    {a b : M} (ha : a ∈ Submodule.span R s)
+    (hb : b ∈ Submodule.span R s) : p a b ha hb := by
```

Comment:
To me it seems a little consistent that the induction principle / motive takes `a b ha hb`, but the `add` and `smul` induction hypotheses takes `a ha b hb`. Maybe this is too annoying to change in this PR, but it would be make them consistent one way or the other eventually.

I slightly favor the `a b ha hb` ordering, since in context this is used as `a b ha hb iha ihb` which is a little more consistent than `a ha b hb iha ihb`.

Reply:
oof, yeah, that's going to be an annoying change, but I can do it if we agree on the ordering. Are you okay with `a b ha hb iha ihb`?